doozer/doozerlib/distgit.py

```diff
@@ -1606,15 +1613,57 @@ def _mapped_image_from_stream(self, image, original_parent, dfp):
         else:
             return parent
 
-    def _rebase_from_directives(self, dfp):
-        self.should_match_upstream = self._should_match_upstream()
+    def _determine_upstream_rhel_version(self, dfp: DockerfileParser) -> Optional[str]:
+        version = None
+
+        try:
+            # We will infer the rhel version from the last build layer in the upstream Dockerfile
+            last_layer = dfp.parent_images[-1]
+            cmd = ['oc', 'image', 'info', '-o', 'json', last_layer]
+            out, _ = exectools.cmd_assert(cmd)
+            tags = json.loads(out)['config']['config']['Labels']['io.openshift.tags']
+
+            # tags will be something like 'base rhel8'
+            pattern = r'el(\d)'
+            match = re.search(pattern, tags)
+            if match:
+                version = match.group()
+
+        except (KeyError, ChildProcessError) as e:
+            # We could get:
+            #   - a ChildProcessError when the upstream equivalent is not found
+            #   - a KeyError when 'io.openshift.tags' label is undefined
+            self.logger.warning('Failed determining upstream rhel version: %s', e)
+
+        if not version:
+            self.logger.warning('Could not determine rhel version from upstream %s', self.name)
+        return version
+
+    def _update_image_config(self):
+        """
+        If we're trying to match upstream, check if there's a 'when' clause in image alternative_config field
+        that matches upstream RHEL version. If so, merge the 'when' clause content with the main image config
+        """
+
         if self.should_match_upstream:
-            # ART-8476 override image config with upstream
-            alt_upstream = self.config.get('alternative_upstream', None)
-            if alt_upstream:
-                self.config.update(alt_upstream)
-                del self.config['alternative_upstream']
+            # Check if there is an alternative configuration matching upstream RHEL version
+            alt_configs = self.config.get('alternative_upstream', None)
+            matched = False
+            for alt_config in alt_configs:
+                if alt_config['when'] == self.upstream_intended_el_version:
+                    self.logger.info('Merging %s alternative config to match upstream', self.upstream_intended_el_version)
+                    self.config.update(alt_config)
+                    matched = True
+                    break
+            if not matched:
+                # there's no 'when' clause matching upstream intended RHEL version: fallback to default ART's config
+                self.logger.warning('Could not match upstream because "%s" version is not mapped in alternative_config',
+                                    self.upstream_intended_el_version)
+                self.should_match_upstream = False
```

Comment:
This would mean that if there is no alt_config defined explicitly, then we won't attempt to use what's upstream .. I'm not sure we want to do that. I think we still want images to use their own upstream golang builder and with the recent change attempt to use their member, if builds fail then we fix it with alt config

Reply:
I have removed line `1662`. If there is no alt config defined, we will attempt to use upsteam but the build will most likely break, as golang and rhel versions won't match. We'll get a Slack notification and be able to react